src/app/clusters/thread-border-router-management-server/thread-br-mgmt-server.cpp

```diff
@@ -0,0 +1,385 @@
+    return aEncoder.Encode(interfaceEnabled);
+}
+
+CHIP_ERROR ServerInstance::ReadActiveDatasetTimestamp(AttributeValueEncoder & aEncoder)
+{
+    Thread::OperationalDataset activeDataset;
+    uint64_t activeDatasetTimestamp;
+    if (mDelegate->GetActiveDataset(activeDataset) == CHIP_NO_ERROR &&
+        activeDataset.GetActiveTimestamp(activeDatasetTimestamp) == CHIP_NO_ERROR)
+    {
+        return aEncoder.Encode(DataModel::MakeNullable(activeDatasetTimestamp));
+    }
+    return aEncoder.Encode(DataModel::Nullable<uint64_t>());
+}
+
+CHIP_ERROR ServerInstance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    if (aPath.mClusterId != ThreadBorderRouterManagement::Id)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    if (!mDelegate)
+    {
+        ChipLogError(Zcl, "Thread Border Router Management server not initialized");
+        return CHIP_ERROR_UNINITIALIZED;
+    }
+    CHIP_ERROR status = CHIP_NO_ERROR;
+    switch (aPath.mAttributeId)
+    {
+    case Globals::Attributes::FeatureMap::Id:
+        status = ReadFeatureMap(aEncoder);
+        break;
+    case Attributes::BorderRouterName::Id:
+        status = ReadBorderRouterName(aEncoder);
+        break;
+    case Attributes::BorderAgentID::Id:
+        status = ReadBorderAgentID(aEncoder);
+        break;
+    case Attributes::ThreadVersion::Id:
+        status = ReadThreadVersion(aEncoder);
+        break;
+    case Attributes::InterfaceEnabled::Id:
+        status = ReadInterfaceEnabled(aEncoder);
+        break;
+    case Attributes::ActiveDatasetTimestamp::Id:
+        status = ReadActiveDatasetTimestamp(aEncoder);
+        break;
+    default:
+        break;
+    }
+    return status;
+}
+
+void ServerInstance::CommitSavedBreadcrumb()
+{
+    if (mBreadcrumb.HasValue())
+    {
+        GeneralCommissioning::SetBreadcrumb(mBreadcrumb.Value());
+    }
+    mBreadcrumb.ClearValue();
+}
+
+void ServerInstance::OnActivateDatasetComplete(CHIP_ERROR error)
+{
+    auto commandHandleRef = std::move(mAsyncCommandHandle);
+    auto commandHandle    = commandHandleRef.Get();
+    if (commandHandle == nullptr)
+    {
+        return;
+    }
+    if (error == CHIP_NO_ERROR)
+    {
+        // The successful completion of the activation process SHALL disarm the fail-safe timer.
+        auto & failSafeContext = chip::Server::GetInstance().GetFailSafeContext();
+        failSafeContext.DisarmFailSafe();
+        CommitSavedBreadcrumb();
+    }
+    else
+    {
+        ChipLogError(Zcl, "Failed on activating the active dataset for Thread BR: %" CHIP_ERROR_FORMAT, error.Format());
+    }
+    commandHandle->AddStatus(mPath, error == CHIP_NO_ERROR ? Status::Success : Status::Failure);
+}
+
+void ServerInstance::OnFailSafeTimerExpired()
+{
+    auto commandHandleRef = std::move(mAsyncCommandHandle);
+    auto commandHandle    = commandHandleRef.Get();
+    if (commandHandle == nullptr)
+    {
+        return;
```

Comment:
This does not look right to me.  Say we got HandleSetActiveDatasetRequest.  We kicked off the async process.  That process finished, and OnActivateDatasetComplete got called with an error.  That nulled out mAsyncCommandHandle, right?  So here we will return without reverting.

At the very least, this needs some explanation in comments about how this stuff is supposed to work together and what the various state bits mean.

Reply:
Moved the `RevertActiveDataset()` before the check of async command handle.